pkg/ndm-grpc/server/services/listBlockDevice.go

```diff
@@ -0,0 +1,281 @@
+/*
+Copyright 2019 The OpenEBS Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"strings"
+
+	"github.com/openebs/node-disk-manager/pkg/apis/openebs/v1alpha1"
+	protos "github.com/openebs/node-disk-manager/pkg/ndm-grpc/protos/ndm"
+	"k8s.io/klog"
+
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/openebs/node-disk-manager/cmd/ndm_daemonset/controller"
+
+	"github.com/openebs/node-disk-manager/pkg/hierarchy"
+	"github.com/openebs/node-disk-manager/pkg/ndm-grpc/server"
+)
+
+// Node helps in using types defined in Server
+type Node struct {
+	server.Node
+}
+
+// NewNode is a constructor
+func NewNode() *Node {
+	return &Node{server.Node{}}
+}
+
+type disks []protos.BlockDevice
+
+//AllBlockDevices contains all the relationships and device types
+type AllBlockDevices struct {
+	Parents    []string
+	Partitions []string
+	LVMs       []string
+	RAIDs      []string
+	Holders    []string
+	Slaves     []string
+	Loops      []string
+	Sparse     []string
+}
+
+var all AllBlockDevices // This variable would contain all devices found in the node and their relationships
+
+// ListBlockDevices returns the block devices and their relationships
+func (n *Node) ListBlockDevices(ctx context.Context, null *protos.Null) (*protos.BlockDevices, error) {
+	klog.Info("Listing block devices")
+
+	ctrl, err := controller.NewController()
+	if err != nil {
+		klog.Errorf("Error creating a controller %v", err)
+		return nil, status.Errorf(codes.NotFound, "Namespace not found")
+	}
+
+	err = ctrl.SetControllerOptions(controller.NDMOptions{ConfigFilePath: "/host/node-disk-manager.config"})
+	if err != nil {
+		klog.Errorf("Error setting config to controller %v", err)
+		return nil, status.Errorf(codes.Internal, "Error setting config to controller")
+	}
+
+	blockDeviceList, err := ctrl.ListBlockDeviceResource(false)
+	if err != nil {
+		klog.Errorf("Error listing block devices %v", err)
+		return nil, status.Errorf(codes.Internal, "Error fetching list of disks")
+	}
+
+	if len(blockDeviceList.Items) == 0 {
+		klog.Info("No items found")
+	}
+
+	blockDevices := make([]*protos.BlockDevice, 0)
+
+	err = GetAllTypes(blockDeviceList)
+	if err != nil {
+		klog.Errorf("Error fetching Parent disks %v", err)
+	}
+
+	for _, name := range all.Parents {
+
+		blockDevices = append(blockDevices, &protos.BlockDevice{
+			Name:       name,
+			Type:       "Disk",
+			Partitions: FilterParitions(name, all.Partitions),
```

Comment:
Why are we taking all partitions and then filtering it? We can get Partitions for single device right?

Reply:
It is to minimize the calls made to `GetDependents`. All the partitions found on a node are stored in  `all.Partitions`.   `GetDependents` is called on each block device only when necessary. Later, while sending the response, we can just filter accordingly based on the block device.